DRAFT: WAM preview for console app

v-michaelmi opened the pull request:
## Purpose
<!-- Describe the intention of the changes being proposed. What problem does it solve or functionality does it add? -->
This is meant to be a sample PR for using WAM instead of IWA. Will likely move to its own REPO soon.

bgavrilMS commented on iwa-console/MSAL/MSALClientHelper.cs:
```diff
@@ -0,0 +1,270 @@
+        /// Attaches the token cache to the Public Client app.
+        /// </summary>
+        /// <returns>IAccount list of already signed-in users (if available)</returns>
+        private async Task<IEnumerable<IAccount>> AttachTokenCache()
```
This method is doing too much, consider using just "AttachToCache"

bgavrilMS commented on iwa-console/appsettings.json:
```diff
@@ -1,28 +1,14 @@
+    "Authority": "https://login.microsoftonline.com/{0}",
+    "TenantId": "[Enter Tenant Id (Obtained from the Azure portal. Select 'Endpoints' from the 'App registrations' blade and use the GUID in any of the URLs), e.g. da41245a5-11b3-996c-00a8-4d99re19f292]",
+    "ClientId": "[Enter the Client Id of the service (Application ID obtained from the Azure portal), e.g. ba74781c2-53c2-442a-97c2-3d60re42f403]",
+    "RedirectURI": "ms-appx-web://microsoft.aad.brokerplugin/{0}",
```
The redirect uri should be "http://localhost" because this is what MSAL.NET uses on Mac and Linux and old Windows. 
`"ms-appx-web://microsoft.aad.brokerplugin/{0}"` only needs to be set at app registration level.

bgavrilMS commented on iwa-console/iwa-console.csproj:
```diff
@@ -2,24 +2,32 @@
+    <PackageReference Include="Microsoft.Identity.Client" Version="4.48.1" />
+    <PackageReference Include="Microsoft.Identity.Client.Broker" Version="4.48.1-preview" />
+    <PackageReference Include="Microsoft.Identity.Client.Extensions.Msal" Version="2.25.0" />
+    <PackageReference Include="Microsoft.Windows.CppWinRT" Version="2.0.221121.5" />
```
Why is this needed?

v-michaelmi replied:
It's not. That was an accidental addition.

bgavrilMS commented on iwa-console/iwa-console.csproj:
```diff
@@ -2,24 +2,32 @@
+    <PackageReference Include="Microsoft.Identity.Client.Broker" Version="4.48.1-preview" />
+    <PackageReference Include="Microsoft.Identity.Client.Extensions.Msal" Version="2.25.0" />
+    <PackageReference Include="Microsoft.Windows.CppWinRT" Version="2.0.221121.5" />
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.2" />
```
Do not use newtonsoft anymore. Use System.Text.Json as per .NET guidelines - https://learn.microsoft.com/en-us/dotnet/api/system.text.json?view=net-7.0

bgavrilMS commented on iwa-console/Program.cs:
```diff
@@ -33,14 +35,32 @@ static void Main(string[] args)
-
-            MyInformation myInformation = new MyInformation(app, httpClient, config.MicrosoftGraphBaseEndpoint);
-            await myInformation.DisplayMeAndMyManagerAsync();
+            // Using appsettings.json as our configuration settings and utilizing IOptions pattern - https://learn.microsoft.com/dotnet/core/extensions/options
```
Please move all this logic into Main and use

`public static Task Main()`

bgavrilMS commented on iwa-console/MSAL/MSALClientHelper.cs:
```diff
@@ -0,0 +1,270 @@
+        /// <value>
+        ///   <c>true</c> if this instance is broker initialized; otherwise, <c>false</c>.
+        /// </value>
+        public bool IsBrokerInitialized { get; private set; }
```
I don't think we should expose this to end-users

v-michaelmi replied:
That can be refactored but are we supporting Brokers for `MacOS` and `Linux`?

bgavrilMS replied:
No, but MSAL will fallback to a browser on Mac / Linux automatically. App developer doesn't need to worry about this.

bgavrilMS commented on iwa-console/MSAL/MSALClientHelper.cs:
```diff
@@ -0,0 +1,270 @@
+        /// Initializes the public client application of MSAL.NET with the required information to correctly authenticate the user.
+        /// </summary>
+        /// <returns>An IAccount of an already signed-in user (if available)</returns>
+        public async Task<IAccount> InitializePublicClientAppForWAMBrokerAsync(IntPtr? handle)
```
Can you just extract the handle value here instead of injecting it from somewhere else?

bgavrilMS commented on iwa-console/MSAL/MSALClientHelper.cs:
```diff
@@ -0,0 +1,270 @@
+            this.IsBrokerInitialized = true;
+
+            await AttachTokenCache();
+            return await FetchSignedInUserFromCache().ConfigureAwait(false);
```
Initialize should not do more than its name. It should not fetch users.

bgavrilMS commented on iwa-console/MSAL/MSALClientHelper.cs:
```diff
@@ -0,0 +1,270 @@
+                    .WithParentActivityOrWindow(() => { return handle.Value; })                                     // Specify Window handle - (required for WAM).
+                    .Build();
+
+            this.IsBrokerInitialized = true;
```
This is bad design, this flag can be set depending on the `Intialize` method used. Please remove flag and `InitializePublicClientAppAsync`

v-michaelmi replied:
Based on what I see there is no signature that only takes an `IntPtr`. Needs to be wrapped in a function.

bgavrilMS commented on iwa-console/MSAL/MSALClientHelper.cs:
```diff
@@ -0,0 +1,270 @@
+            // Initialize the MSAL library by building a public client application for authenticating using WAM
+            this.PublicClientApplication = this.PublicClientApplicationBuilder
+                    .WithBrokerPreview(true)
+                    .WithParentActivityOrWindow(() => { return handle.Value; })                                     // Specify Window handle - (required for WAM).
```
Set this only on Windows

bgavrilMS commented on iwa-console/MSAL/MSALClientHelper.cs:
```diff
@@ -0,0 +1,270 @@
+            // Initialize the MSAL library by building a public client application for authenticating using WAM
+            this.PublicClientApplication = this.PublicClientApplicationBuilder
+                    .WithBrokerPreview(true)
+                    .WithParentActivityOrWindow(() => { return handle.Value; })                                     // Specify Window handle - (required for WAM).
```
No need to use Func or .Value here. Just pass in the IntPtr i.e. `WithParentActivityOrWindow(handle)`

v-michaelmi replied:
Based on what I see there is no signature that only takes an IntPtr. Needs to be wrapped in a function.

bgavrilMS replied:
Ah ok yeah, you are right. That version is on AcquireTokenInteractive API. This is fine.

bgavrilMS commented on iwa-console/MSAL/MSALClientHelper.cs:
```diff
@@ -0,0 +1,270 @@
+                }
+                else
+                {
+                    if (this.IsBrokerInitialized)
```
Broker only please.

bgavrilMS commented on iwa-console/MSAL/MSALClientHelper.cs:
```diff
@@ -0,0 +1,270 @@
+                    }
+                    else
+                    {
+                        this.AuthResult = await SignInUserInteractivelyAsync(scopes);
```
Won't be needed in broker only, since `AcquireTokenSilent` will throw `MsalUiRequiredException `

v-michaelmi commented:
@bgavrilMS 

Addressed the feedback you have mentioned. I updated the sample to use **NET7** and seems to work fine.

Only snag is the `WithParentActivityOrWindow` method only has signatures that accept `Func` parameters. Nothing seems to take a standalone `IntPtr`

v-michaelmi commented:
Also added a trivial sign-out flow

v-michaelmi commented:
@bgavrilMS running into a snag running this on Linux. If I use the `WithBroker` option it fails, and I don't believe we have support for Linux do we?

When I disable it, it works fine along with caching.

bgavrilMS commented:
> @bgavrilMS running into a snag running this on Linux. If I use the `WithBroker` option it fails, and I don't believe we have support for Linux do we?
> 
> When I disable it, it works fine along with caching.

Sorry, my bad, that API with IntPtr is on AcquireTokenInteractive. This API with Func<IntPtr> is on PCA object. This is fine./

v-michaelmi commented:
@bgavrilMS That should do it. Works fine when I test it on Linux.

I have this setup so that it assumes the user wants to use **WAM** when on a windows device but let me know and I can add a flag for the user to choose if they want to use a browser instead.

If this LGTY I can make a repository or just rework what we have here. Whichever's easiest.

v-michaelmi commented:
Obsolete.

Closing.
